### PR TITLE
feat: Introduce NotFoundFailureView

### PR DESCRIPTION
**Pull request by olerichter00**
This PR resolves [CX-2456]

Depends on https://github.com/artsy/echo/pull/251

### Description

This PR introduces a new "Not Found" error view with a "go back" button, that indicates that a resource couldn’t be found (e.g. an artwork that has been deleted). The “not found” text can be customized and falls back to a generic “not found” text.

The new view is shown to the user whenever the HTTP error code 404 is present in an error response:

```
error?.res?.json?.errors?.[0]?.extensions?.httpStatusCodes.includes(404)
```

| My Collection Artwork Screen | Generic  |
| --- | --- |
|Simulator Screen Shot - iPhone 12 Pro Max - 2022-05-05 at 11 33 43 |Simulator Screen Shot - iPhone 12 Pro Max - 2022-05-05 at 11 35 17 |

### PR Checklist

- [ ] I tested my changes on **iOS** / **Android**.
- [ ] I added screenshots or videos to illustrate my changes.
- [ ] I added Tests and Stories for my changes.
- [ ] I added an [app state migration].
- [ ] I hid my changes behind a [feature flag].

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to run this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

- Introduce "Not Found" error view with back button - ole

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

</details>

Need help with something? Have a look on our [docs], or get in touch with us.

[app state migration]: /docs/adding_state_migrations.md
[feature flag]: /docs/developing_a_feature.md
[docs]: /docs/README.md

[CX-2456]: https://artsyproduct.atlassian.net/browse/CX-2456?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/app/Components/RetryErrorBoundary.tsx [hunk 1]
```diff
@@ -67,13 +72,31 @@ export class RetryErrorBoundary extends Component<
...
 
     if (error) {
       if (failureView) {
         return failureView({ error, retry: this._retry })
       }
+
+      // @ts-expect-error
+      const httpStatusCodes = error?.res?.json?.errors?.[0]?.extensions?.httpStatusCodes || []
+
+      const isNotFoundError = httpStatusCodes.includes(404)
```
Comment (MounirDhahri) [inline src/app/Components/RetryErrorBoundary.tsx:87]: **thought:** the 404 here only occurs when the asset is missing, right?

File: src/app/Components/RetryErrorBoundary.tsx [hunk 2]
```diff
@@ -67,13 +72,31 @@ export class RetryErrorBoundary extends Component<
...
+      this.props
     const { error } = this.state
 
     if (error) {
       if (failureView) {
         return failureView({ error, retry: this._retry })
       }
+
+      // @ts-expect-error
+      const httpStatusCodes = error?.res?.json?.errors?.[0]?.extensions?.httpStatusCodes || []
```
Comment (MounirDhahri) [inline src/app/Components/RetryErrorBoundary.tsx:85]: **thought:** would it make sense as well to check for this, only when it's happening on a query that doesn't contain the `optionalField` directive as we do for the `errorMiddleware` ?

File: src/app/Components/RetryErrorBoundary.tsx (hunk 1 above)
Comment (olerichter00) [reply src/app/Components/RetryErrorBoundary.tsx:87]: It occurs for example if an artwork is missing in Gravity.

File: src/app/Components/RetryErrorBoundary.tsx (hunk 2 above)
Comment (olerichter00) [reply src/app/Components/RetryErrorBoundary.tsx:85]: If the error happens inside of an optional field, the `errorMiddleware` should catch the error anyway.

**Review by brainbicycle — DISMISSED** [review]
love it! some minor comments otherwise looks great

**Pull request merged**